// @storybook/react-native/dist/preview/components/OnDeviceUI/navigation/Navigation.js
import React from 'react';
import { View } from 'react-native';
import { useSafeAreaInsets } from 'react-native-safe-area-context';
import GestureRecognizer from 'react-native-swipe-gestures';
import Bar from './Bar';
import VisibilityButton from './VisibilityButton';
const SWIPE_CONFIG = {
    velocityThreshold: 0.2,
    directionalOffsetThreshold: 80,
};
const navStyle = {
    position: 'absolute',
    left: 0,
    right: 0,
    bottom: 0,
};
const Navigation = ({ tabOpen, onChangeTab, isUIVisible, setIsUIVisible }) => {
    const insets = useSafeAreaInsets();
    const handleToggleUI = () => {
        setIsUIVisible((oldIsUIVisible) => !oldIsUIVisible);
    };
    const handleSwipeLeft = () => {
        if (tabOpen < 1) {
            onChangeTab(tabOpen + 1);
        }
    };
    const handleSwipeRight = () => {
        if (tabOpen > -1) {
            onChangeTab(tabOpen - 1);
        }
    };
    return (React.createElement(View, { style: navStyle },
        isUIVisible && (React.createElement(GestureRecognizer, { onSwipeLeft: handleSwipeLeft, onSwipeRight: handleSwipeRight, config: SWIPE_CONFIG },
            React.createElement(Bar, { index: tabOpen, onPress: onChangeTab, style: { paddingBottom: insets.bottom } }))),
        React.createElement(VisibilityButton, { onPress: handleToggleUI, style: { marginBottom: insets.bottom } })));
};
export default React.memo(Navigation);
