// @storybook/addon-ondevice-controls/dist/GroupTabs.js
import React, { Component } from 'react';
import { ScrollView, StyleSheet, Text, TouchableOpacity } from 'react-native';
import styled from '@emotion/native';
const Label = styled.Text(({ theme, active }) => ({
    color: active ? theme.buttonActiveTextColor || '#444444' : theme.buttonTextColor || '#999999',
    fontSize: 17,
}));
class GroupTabs extends Component {
    renderTab(name, group) {
        let { title } = group;
        if (typeof title === 'function') {
            title = title();
        }
        const { onGroupSelect, selectedGroup } = this.props;
        return (React.createElement(TouchableOpacity, { style: styles.tab, key: name, onPress: () => onGroupSelect(name) },
            React.createElement(Label, { active: selectedGroup === name }, title)));
    }
    render() {
        const { groups } = this.props;
        const entries = groups ? Object.entries(groups) : null;
        return entries && entries.length ? (React.createElement(ScrollView, { horizontal: true, style: styles.scrollView }, entries.map(([key, value]) => this.renderTab(key, value)))) : (React.createElement(Text, null, "no groups available"));
    }
}
GroupTabs.defaultProps = {
    groups: {},
    onGroupSelect: () => { },
    selectedGroup: null,
};
const styles = StyleSheet.create({
    scrollView: {
        marginHorizontal: 10,
        flexDirection: 'row',
        marginBottom: 10,
        borderBottomWidth: 1,
        borderBottomColor: '#ccc',
    },
    tab: {
        marginTop: 5,
        marginRight: 15,
        paddingBottom: 10,
    },
});
export default GroupTabs;
